app/src/main/java/atorch/shortestpaths/SolvePuzzle.java

```diff
@@ -31,264 +31,264 @@
+                current_path_tokenized = current_path.split(";");
+                for (int j = 0; j < path_length_inner; j++)
+                    paths[i][j] = countries[Integer.parseInt(current_path_tokenized[j])];
+                i++;
+            }
+            c2.close();
+
+            setPuzzleStatement(res);
+
+            Button buttonGiveUp = (Button) findViewById(R.id.button_give_up);
+            buttonGiveUp.setOnClickListener(new OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    openGiveUpDialog(v);
+                }
+            });
+
+            final AutoCompleteTextView country_input = (AutoCompleteTextView) findViewById(R.id.autocomplete_country);
+
+            final ArrayAdapter<String> adapter = new ArrayAdapter<String>(this, android.R.layout.simple_list_item_1, countries);
+            country_input.setAdapter(adapter);
+            country_input.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+                @Override
+                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+                    String user_answer = adapter.getItem(position).toString();
+                    boolean user_answer_is_correct = false;
+                    boolean new_path_indicators[] = new boolean[number_of_paths];  // Default to false
+                    for (int i = 0; i < number_of_paths; i++) {
+                        if (path_indicators[i] && paths[i][user_progress] == user_answer) {
+                            user_answer_is_correct = true;
+                            new_path_indicators[i] = true;
+                        }
+                    }
+                    if (user_answer_is_correct) {
+                        user_progress++;
+                        path_indicators = new_path_indicators;
+                        addCountryVisited(context, user_answer);
+                        if (user_progress < path_length_inner) {
+                            String congratulations = congratulations_array[random.nextInt(congratulations_array.length)];
+                            Toast.makeText(context, congratulations, Toast.LENGTH_SHORT).show();
+                        } else {
+                            // Done solving puzzle: make country_input invisible; close keyboard
+                            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
+                            imm.hideSoftInputFromWindow(country_input.getWindowToken(), 0);
+                            country_input.setVisibility(View.GONE);
+
+                            addDestinationCountry(context, res);
+
+                            SharedPreferences sharedPref = getSharedPreferences("atorch.shortestpaths.data", Context.MODE_PRIVATE);
+                            int new_points = level + path_length_inner;
```

Comment:
This and the lines below will have to be updated: instead of adding points based on the path length, we have to count how many _new_ (i.e. previously unseen) countries were visited